### PR TITLE
[SPARK-49792][PYTHON][BUILD] Upgrade to numpy 2 for building and testing Spark branches

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Upgrade numpy to 2.1.0 for building and testing Spark branches.

Failed tests are categorized into the following groups:
- Most of test failures fixed are related to https://github.com/pandas-dev/pandas/issues/59838#event-14332587978.
- Replaced np.mat with np.asmatrix.
- TODO: SPARK-49793

### Why are the changes needed?
Ensure compatibility with newer NumPy, which is utilized by Pandas (on Spark).

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by bjornjorgensen** [comment]
there is a new panda version https://pandas.pydata.org/pandas-docs/version/2.2.3/whatsnew/v2.2.3.html that have support for numpy 2.1https://github.com/pandas-dev/pandas/pull/59444

**Comment by xinrong-meng** [comment]
```
[info] - SPARK-47148: AQE should avoid to submit shuffle job on cancellation *** FAILED *** (6 seconds, 99 milliseconds)
[info]   "[SCALAR_SUBQUERY_TOO_MANY_ROWS] More than one row returned by a subquery used as an expression. SQLSTATE: 21000
[info]   == SQL (line 1, position 12) ==
[info]   SELECT id, (SELECT slow_udf() FROM range(2)) FROM range(5)
[info]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[info]   " did not contain "coalesce test error" (AdaptiveQueryExecSuite.scala:939)
```
seems irrelevant. Retriggered.

**Comment by xinrong-meng** [comment]
Thank you @bjornjorgensen!
I think we can separate the pandas upgrade from the numpy upgrade, as the current pandas version should be compatible with numpy 2.1.0 as well.

File: python/pyspark/ml/tests/test_functions.py [hunk 1]
```diff
@@ -193,6 +195,10 @@ def predict(inputs):
         batch_sizes = preds["preds"].to_numpy()
         self.assertTrue(all(batch_sizes <= batch_size))
 
+    # TODO(SPARK-49793): enable the test below
```
Comment (xinrong-meng) [inline python/pyspark/ml/tests/test_functions.py:198]: @WeichenXu123 may I get your input on that please?
More details can be found here https://issues.apache.org/jira/browse/SPARK-49793.

**Comment by codesorcery** [comment]
Maybe helpful here: the Ruff linter/formatter has some rules to check for NumPy 2 deprecations (https://docs.astral.sh/ruff/rules/numpy2-deprecation/).
I intended to create a pull request for adding those checks to the build pipeline after https://github.com/apache/spark/pull/47083 was merged, but unfortunately didn't find the time back then.

**Comment by xinrong-meng** [comment]
The test failures we are trying to fix here are almost all related to [this issue](https://github.com/pandas-dev/pandas/issues/59838#event-14332587978). Thank you @codesorcery for sharing!

**Comment by xinrong-meng** [comment]
Retriggered irrelevant tests

**Comment by xinrong-meng** [comment]
```
[info] - interrupt all - background queries, foreground interrupt *** FAILED *** (20 seconds, 50 milliseconds)
[info]   The code passed to eventually never returned normally. Attempted 30 times over 20.046569918 seconds. Last failure message: q2Interrupted was false. (SparkSessionE2ESuite.scala:71)
[info]   org.scalatest.exceptions.TestFailedDueToTimeoutException:
```
Retriggering

File: python/pyspark/ml/tests/test_functions.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/ml/tests/test_functions.py:197]: do you have error message and error stack for numpy2 + caching ?

Comment (xinrong-meng) [reply python/pyspark/ml/tests/test_functions.py:197]: Would you please see https://issues.apache.org/jira/browse/SPARK-49793? There is no error but the results are unexpected.

**Comment by xinrong-meng** [comment]
@HyukjinKwon @zhengruifeng @dongjoon-hyun would you please review?

**Comment by zhengruifeng** [comment]
In General LGTM, pending @WeichenXu123 's feedback on the failed ml caching test

File: python/pyspark/ml/tests/test_functions.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/ml/tests/test_functions.py:197]: Got it , need some time to investigation, but we can disable it as a workaround for now.

Comment (xinrong-meng) [reply python/pyspark/ml/tests/test_functions.py:197]: Sounds good, thank you!

Comment (zhengruifeng) [reply python/pyspark/ml/tests/test_functions.py:197]: also cc @leewyang as the test author

**Review by zhengruifeng — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @xinrong-meng and all.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0 on February 2025.

**Comment by xinrong-meng** [comment]
Thank you @dongjoon-hyun !

**Pull request closed without merging**